### PR TITLE
[msbuild/dotnet] Don't use the built-in publishing logic in .NET to copy frameworks to the app bundle. Fixes #12369.

### LINKED ISSUE #12369: [maccatalyst] NativeReference results in "bundle format is ambiguous (could be app or framework)"
### Steps to Reproduce

1. Create a new net6.0-maccatalyst app
2. Add a `<NativeReference>` to a Mac Catalyst dynamic .framework
3. Observe a compiler error: `bundle format is ambiguous (could be app or framework)`

I did notice that my framework has symlinks, but once it reaches the .app it has lost the symlinks and has duplicate files.

### Expected Behavior

App builds and links.

### Actual Behavior

```
/usr/local/share/dotnet/packs/Microsoft.MacCatalyst.Sdk/15.0.100-preview.7227/tools/msbuild/iOS/Xamarin.Shared.targets(1140,3): 
error : bin/Debug/net6.0-maccatalyst/maccatalyst-x64/SkiaSharpTestAppMacCat.app/Contents/Frameworks/libSkiaSharp.framework: bundle format is ambiguous (could be app or framework)
```

### Environment

```
dotnet --info
.NET SDK (reflecting any global.json):
 Version:   6.0.100-preview.7.21379.14
 Commit:    22d70b47bc

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  11.5
 OS Platform: Darwin
 RID:         osx.11.0-x64
 Base Path:   /usr/local/share/dotnet/sdk/6.0.100-preview.7.21379.14/

Host (useful for support):
  Version: 6.0.0-preview.7.21377.19
  Commit:  91ba01788d

.NET SDKs installed:
  3.1.411 [/usr/local/share/dotnet/sdk]
  5.0.302 [/usr/local/share/dotnet/sdk]
  6.0.100-preview.7.21377.3 [/usr/local/share/dotnet/sdk]
  6.0.100-preview.7.21377.35 [/usr/local/share/dotnet/sdk]
  6.0.100-preview.7.21379.14 [/usr/local/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 3.1.17 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.8 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0-preview.7.21376.6 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0-preview.7.21377.13 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0-preview.7.21378.6 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0-rc.1.21365.18 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.17 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.8 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0-preview.7.21377.12 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0-preview.7.21377.19 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download

```

### Build Logs

[msbuild.binlog.zip](https://github.com/xamarin/xamarin-macios/files/6943202/msbuild.binlog.zip)

### Example Project (If Possible)

[SkiaSharpTestAppMacCat.zip](https://github.com/xamarin/xamarin-macios/files/6943204/SkiaSharpTestAppMacCat.zip)

Extract the project and then unzip `libSkiaSharp.framework.zip` making sure to preserve symlinks.

### PR DESCRIPTION
**Pull request by rolfbjarne**
.NET/MSBuild don't handle symlinks properly [1], which means that we can't ask
.NET to copy frameworks to the app bundle, since frameworks may contain
symlinks.

In our case, the symptom was that instead of copying symlinks, the file the
symlink pointed to was copied instead, and then codesign complained about
invalid bundle format when we tried to sign the framework.

We fix this by having our own target (_CopyFrameworksToBundle) to copy
frameworks to the app bundle (instead of adding all the files in the
frameworks to the ResolvedFileToPublish item group), and then using 'ditto' to
copy the frameworks.

In order to create a test case for this, I also made the macOS and Mac
Catalyst versions of the XTest framework use symlinks:

* Create a proper XTest framework bundle hierarchy for macOS and Mac Catalyst
  by using the typical symlink structure (actual files in the Versions/A
  subdirectory, and then symlinks pointing into that directory).
* Create a separate Info.plist for each platform for XTest.framework, since
  using an otherwise correct framework makes tooling (such as codesign)
  complain if the Info.plist isn't correct too.

This made our existing tests show the bug.

Finally I had to fix signing frameworks where the executable is a symlink.

We were first resolving symlinks for the input - say we had an
Example.framework/Example symlink to Example.framework/Versions/A/Example -
and then checking the parent directory if it's a framework. The parent
directory of 'Example.framework/Versions/A/Example' is 'A', which did not meet
our framewrok condition (if it ends with '.framework').

The fix is to adjust the logic to resolve symlinks after checking if the input
is a framework or not.

[1]: https://github.com/dotnet/msbuild/issues/6821

Fixes https://github.com/xamarin/xamarin-macios/issues/12369.

### CODE REVIEW

**Review by spouliot — APPROVED** [review]

**Comment by spouliot** [comment]
build failed with
```
error: unable to read the file at '/Users/builder/azdo/_work/1/s/xamarin-macios/tests/test-libraries/.libs/macos/XTest.framework/XTest'
make[2]: *** [.libs/XTest.xcframework] Error 70
```

**Review by chamons — APPROVED** [review]

File: dotnet/targets/Xamarin.Shared.Sdk.targets [hunk 1]
```diff
@@ -560,30 +561,39 @@
...
-			<ResolvedFileToPublish Include="@(_FrameworkFilesToPublish)">
-				<RelativePath>%(_FrameworkFilesToPublish._FrameworkPath)\%(_FrameworkFilesToPublish._FrameworkRelativePath)</RelativePath>
-				<CopyToPublishDirectory>PreserveNewest</CopyToPublishDirectory>
-			</ResolvedFileToPublish>
+			<!-- Collect the list of frameworks to publish from _FrameworkNativeReference. The ExtractBindingLibrariesStep in the linker might also add frameworks to _FrameworkToPublish -->
+			<_FrameworkToPublish Include="@(_FrameworkNativeReference)" Condition="'%(_FrameworkNativeReference.Kind)' == 'Framework' And '%(_FrameworkNativeReference.CopyToAppBundle)' != 'false'" />
+
+			<!-- Set TargetDirectory and SourceDirectory for all frameworks we have to publish -->
+			<_FrameworkToPublish Update="@(_FrameworkToPublish)">
+				<TargetDirectory>$(PublishDir)$(_AppBundleFrameworksDir)\%(Filename).framework</TargetDirectory>
```
Comment (emaf) [inline dotnet/targets/Xamarin.Shared.Sdk.targets:572]: Is `PublishDir` a relative path? If not we may end up using Windows paths on the Mac.

Comment (rolfbjarne) [reply dotnet/targets/Xamarin.Shared.Sdk.targets:572]: It's relative by default, but customers might set it to a full path (we don't change it).

I'll change it to make sure we're using a relative path in either case.

Comment (rolfbjarne) [reply dotnet/targets/Xamarin.Shared.Sdk.targets]: @emaf I've fixed it now to use a relative path

**Review by emaf — APPROVED** [review]

**Comment by hamarb123** [comment]
This looks good!
Please ping me when it gets merged - Thanks.

**Comment by rolfbjarne** [comment]
Test failure is unrelated (https://github.com/xamarin/maccore/issues/2414).

**Pull request merged**